pkg/execution/taskgroup.go

```diff
@@ -0,0 +1,121 @@
+package execution
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// A TaskGroup manages a collection of cooperating goroutines. Add new tasks to the group with the Run method.
+// Call the Wait method to wait for the tasks to complete.
+// A zero value is ready for use, but must not be copied after its first use.
+//
+// The group collects any errors returned by the tasks in the group.
+// The first non-nil error reported by any execution and not filtered is returned from the Wait method.
+type TaskGroup struct {
+	wg sync.WaitGroup // Counter for active goroutines.
+
+	// active is nonzero when the group is "active", meaning there has been at least one call to Run since the group
+	// was created or the last Wait.
+	//
+	// Together active and errLock work as a kind of resettable sync.Once. The fast path reads active and only
+	// acquires errLock if it discovers setup is needed.
+	active atomic.Uint32
+
+	errLock sync.Mutex // Guards the fields below.
+	err     error      // First captured error returned from Wait.
```

Comment:
The `errgroup` is also reusable. But the main problem with `errgroup` is that it terminates all controlled gorutines on first error in any goroutine. The `taskgroup` works differently, it waits for all goroutines to complete and collect errors by joining them into one. 
Or you ask about something else?

Reply:
> is that it terminates all controlled gorutines on first error in any goroutine.

Nope, it's not. It just cancels the connected context, but terminates nothing.
About errors. As I can understand (maybe wrong) `TaskGroup` saves only the first non-nil error same as an `errgoup` does.

<details><summary>Suggested diff</summary>
<p>

```diff
diff --git a/pkg/execution/taskgroup.go b/pkg/execution/taskgroup.go
index 387d4e00..e7dd9df0 100644
--- a/pkg/execution/taskgroup.go
+++ b/pkg/execution/taskgroup.go
@@ -3,6 +3,8 @@ package execution
 import (
 	"sync"
 	"sync/atomic"
+
+	"golang.org/x/sync/errgroup"
 )
 
 // A TaskGroup manages a collection of cooperating goroutines. Add new tasks to the group with the Run method.
@@ -12,8 +14,7 @@ import (
 // The group collects any errors returned by the tasks in the group.
 // The first non-nil error reported by any execution and not filtered is returned from the Wait method.
 type TaskGroup struct {
-	wg sync.WaitGroup // Counter for active goroutines.
-
+	eg errgroup.Group // The underlying errgroup.
 	// active is nonzero when the group is "active", meaning there has been at least one call to Run since the group
 	// was created or the last Wait.
 	//
@@ -22,7 +23,6 @@ type TaskGroup struct {
 	active atomic.Uint32
 
 	errLock sync.Mutex // Guards the fields below.
-	err     error      // First captured error returned from Wait.
 	onError errorFunc  // Called each time a task returns non-nil error.
 }
 
@@ -55,16 +55,13 @@ func (g *TaskGroup) OnError(handler func(error) error) *TaskGroup {
 // Run starts an [execute] function in a new goroutine in [TaskGroup]. The execution is not interrupted by TaskGroup,
 // so the [execute] function should include the interruption logic.
 func (g *TaskGroup) Run(execute func() error) {
-	g.wg.Add(1)
 	if g.active.Load() == 0 {
 		g.activate()
 	}
-	go func() {
-		defer g.wg.Done()
-		if err := execute(); err != nil {
-			g.handleError(err)
-		}
-	}()
+	g.eg.Go(func() error {
+		err := execute()
+		return g.handleError(err)
+	})
 }
 
 // Wait blocks until all the goroutines currently active in the TaskGroup have returned, and all reported errors have
@@ -77,7 +74,7 @@ func (g *TaskGroup) Run(execute func() error) {
 //
 // Wait may be called from at most one goroutine at a time. After Wait has returned, the group is ready for reuse.
 func (g *TaskGroup) Wait() error {
-	g.wg.Wait()
+	err := g.eg.Wait() // Wait for all tasks to finish, it returns the first non-nil error.
 	g.errLock.Lock()
 	defer g.errLock.Unlock()
 
@@ -85,7 +82,7 @@ func (g *TaskGroup) Wait() error {
 	if g.active.Load() != 0 {
 		g.active.Store(0)
 	}
-	return g.err
+	return err
 }
 
 // activate resets the state of the group and marks it as "active". This is triggered by adding a goroutine to
@@ -94,19 +91,19 @@ func (g *TaskGroup) activate() {
 	g.errLock.Lock()
 	defer g.errLock.Unlock()
 	if g.active.Load() == 0 {
-		g.err = nil
+		g.eg = errgroup.Group{} // Reset the errgroup.
 		g.active.Store(1)
 	}
 }
 
 // handleError synchronizes access to the error handler and captures the first non-nil error.
-func (g *TaskGroup) handleError(err error) {
+func (g *TaskGroup) handleError(err error) error {
+	if err == nil {
+		return nil
+	}
 	g.errLock.Lock()
 	defer g.errLock.Unlock()
-	e := g.onError.filter(err)
-	if e != nil && g.err == nil {
-		g.err = e // Capture the first unfiltered error.
-	}
+	return g.onError.filter(err)
 }
 
 // An errorFunc is called by a group each time an execution reports an error. Its return value replaces the reported

```

</p>
</details>

Reply:
@alexeykiselev Found an error in my previous `Suggested diff` inside the `TaskGroup.activate` method. It's important to check the atomic `activate` variable twice (before the lock and after).

Reply:
I won't use ErrGroup it looks like overkill here. But I've replaced the `atomic.Uint32` with `atomic.Bool` and used `CompareAndSwap` where it's possible.